### PR TITLE
MRG: More detailed error message when trying to write modified data

### LINKED ISSUE #718: Asserting equal times in Raw and Original results in unwarranted failure for (some) unstructured files
Hi, thanks for creating and maintaining this package :) I think I may be stumbling across a "it's a feature!-bug".

**100-feet overview**
I have inherited a badly structured MEG dataset and use mne-bids to create a well-structured version from it. 
The raw file names are not consistent, and split files will thus not be read in consecutively automatically. My scripts read in the raw data split-wise and concatenate them. I don't want to touch/modify/rename the raw data - many people use it, and I want my scripts to be reproducible from the very start.
 
When saving this manually read-in raw file with ``write_raw_bids``, mne-bids  performs an internal check that the ``.times`` of the passed ``Raw`` instance (manually read-in, concatenated file) and the ``.times`` of the "original data" (read in automatically from the internal filepath of the ``Raw`` object) match. As the internal filepath points to a file that will not be read with all consecutive splits automatically, this check fails.

## Example
The files I am reading in look for example like this:
```
1_memento_001_ml83.fif  
2_memento_001_ml83-1.fif  
memento_001_ml83-2.fif
```
or 
```
1_memento_005_dk76.fif
2_memento_005_DK76.fif  
3_memento_005_DK76-1.fif
```

[My code](https://github.com/adswa/pymento/blob/19cc9de3a85a7ecb2287c8d48ae500613bbbb9d4/pymento_meg/orig/restructure.py#L111) reads in the data and eventually calls ``write_raw_bids`` like this:

```py
    bids_path = _get_BIDSPath(subject, bidsdir)
    print(
        f"Saving BIDS-compliant raw data from subject "
        f"{subject} into " f"{bids_path}"
    )
    # save raw fif data and events
    events_data, event_dict = _events(raw, subject, figdir)
    write_raw_bids(raw, bids_path, events_data=events_data,
                   event_id=event_dict, overwrite=True)

    # save crosstalk and calibration file
    _elektas_extras(crosstalk_file=ctfile,
                    fine_cal_file=fcfile,
                    bids_path=bids_path)
```

Within ``write_raw_bids``, this check is causing the operation to fail:

https://github.com/mne-tools/mne-bids/blob/a45695ed43394f067a776f0f51441723915bcbf5/mne_bids/write.py#L1134-L1152

I get why this check exists, but could I interest you in the thought of for example giving it a ``on_fail='ignore'/'warn'/'raise'`` parameter, just like many functions in mne, or a ``be_extra_sure=True/False`` argument for ``write_raw_bids``? (or something similar - whatever it may be, I'd be happy to attempt a PR). To me, being able to override this check would be preferable over the solution of "rename your raw files".

Granted, renaming the raw files would solve it, but this would modify pristine raw data that for many years has been used by dozens of researchers. My scripts wouldn't run on this data, but require a personal copy, and a separate step in my script to first copy and rename the data in a temporary directory would double the amount of required disk space storage.

Cheers,
Adina

<details><summary>Full traceback</summary>

```
pymento-restructure --subject '005' --raw_data_dir Data_MEG/RawData_MEG_memento/ --bids_dir /tmp/anotherbids --calfiles_dir  Data_MEG/RawData_MEG_memento/ --diagnostics_dir /tmp/anotherbidsfigs
Starting to restructure original memento data into BIDS for subject sub-005.
Reading files for subject sub-005 from Data_MEG/RawData_MEG_memento/memento_005/Raw/*.fif.
Order the files as follows: Data_MEG/RawData_MEG_memento/memento_005/Raw/1_memento_005_dk76.fif, Data_MEG/RawData_MEG_memento/memento_005/Raw/2_memento_005_DK76.fif, Data_MEG/RawData_MEG_memento/memento_005/Raw/3_memento_005_DK76-1.fif
Opening raw data file Data_MEG/RawData_MEG_memento/memento_005/Raw/1_memento_005_dk76.fif...
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/orig/restructure.py:86: RuntimeWarning: This filename (Data_MEG/RawData_MEG_memento/memento_005/Raw/1_memento_005_dk76.fif) does not conform to MNE naming conventions. All raw files should end with raw.fif, raw_sss.fif, raw_tsss.fif, raw.fif.gz, raw_sss.fif.gz, raw_tsss.fif.gz or _meg.fif
  split1 = mne.io.read_raw_fif(first, on_split_missing="warn")
    Read a total of 8 projection items:
        grad_68.fif : PCA-v1 (1 x 306)  idle
        grad_68.fif : PCA-v2 (1 x 306)  idle
        grad_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v1 (1 x 306)  idle
        mag_68.fif : PCA-v2 (1 x 306)  idle
        mag_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v4 (1 x 306)  idle
        mag_68.fif : PCA-v5 (1 x 306)  idle
    Range : 863000 ... 1633999 =    863.000 ...  1633.999 secs
Ready.
Opening raw data file Data_MEG/RawData_MEG_memento/memento_005/Raw/2_memento_005_DK76.fif...
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/orig/restructure.py:87: RuntimeWarning: This filename (Data_MEG/RawData_MEG_memento/memento_005/Raw/2_memento_005_DK76.fif) does not conform to MNE naming conventions. All raw files should end with raw.fif, raw_sss.fif, raw_tsss.fif, raw.fif.gz, raw_sss.fif.gz, raw_tsss.fif.gz or _meg.fif
  split2 = mne.io.read_raw_fif(second, on_split_missing="warn")
    Read a total of 8 projection items:
        grad_68.fif : PCA-v1 (1 x 306)  idle
        grad_68.fif : PCA-v2 (1 x 306)  idle
        grad_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v1 (1 x 306)  idle
        mag_68.fif : PCA-v2 (1 x 306)  idle
        mag_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v4 (1 x 306)  idle
        mag_68.fif : PCA-v5 (1 x 306)  idle
    Range : 41000 ... 1637999 =     41.000 ...  1637.999 secs
Ready.
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/orig/restructure.py:87: RuntimeWarning: Split raw file detected but next file /media/adina/elements/backup/memento/data/DMS_MEMENTO/Data_MEG/RawData_MEG_memento/memento_005/Raw/2_memento_005_DK76-1.fif does not exist. Ensure all files were transferred properly and that split and original files were not manually renamed on disk (split files should be renamed by loading and re-saving with MNE-Python to preserve proper filename linkage).
  split2 = mne.io.read_raw_fif(second, on_split_missing="warn")
Opening raw data file Data_MEG/RawData_MEG_memento/memento_005/Raw/3_memento_005_DK76-1.fif...
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/orig/restructure.py:88: RuntimeWarning: This filename (Data_MEG/RawData_MEG_memento/memento_005/Raw/3_memento_005_DK76-1.fif) does not conform to MNE naming conventions. All raw files should end with raw.fif, raw_sss.fif, raw_tsss.fif, raw.fif.gz, raw_sss.fif.gz, raw_tsss.fif.gz or _meg.fif
  split3 = mne.io.read_raw_fif(third, on_split_missing="warn")
    Read a total of 8 projection items:
        grad_68.fif : PCA-v1 (1 x 306)  idle
        grad_68.fif : PCA-v2 (1 x 306)  idle
        grad_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v1 (1 x 306)  idle
        mag_68.fif : PCA-v2 (1 x 306)  idle
        mag_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v4 (1 x 306)  idle
        mag_68.fif : PCA-v5 (1 x 306)  idle
    Range : 1638000 ... 3096999 =   1638.000 ...  3096.999 secs
Ready.
/tmp/anotherbids/sub-005
Saving BIDS-compliant raw data from subject 005 into /tmp/anotherbids/sub-005/meg/sub-005_task-memento_meg.fif
Trigger channel has a non-zero initial value of 32768 (consider using initial_event=True to detect this event)
3562 events found
Event IDs: [    2    10    12    13    14    15    16    17    18    19    20    22
    25    26    27    28 32771 32790 32792]
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/utils.py:108: RuntimeWarning: 1 from event_id is not present in events.
  fig = mne.viz.plot_events(
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/utils.py:108: RuntimeWarning: 4 from event_id is not present in events.
  fig = mne.viz.plot_events(
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/utils.py:108: RuntimeWarning: 5 from event_id is not present in events.
  fig = mne.viz.plot_events(
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/utils.py:108: RuntimeWarning: 21 from event_id is not present in events.
  fig = mne.viz.plot_events(
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/utils.py:108: RuntimeWarning: 24 from event_id is not present in events.
  fig = mne.viz.plot_events(
/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/utils.py:108: RuntimeWarning: More events than default colors available. You should pass a list of unique colors.
  fig = mne.viz.plot_events(
/tmp/anotherbidsfigs/sub-005/meg
Opening raw data file Data_MEG/RawData_MEG_memento/memento_005/Raw/1_memento_005_dk76.fif...
/home/adina/repos/mne-bids/mne_bids/write.py:1148: RuntimeWarning: This filename (Data_MEG/RawData_MEG_memento/memento_005/Raw/1_memento_005_dk76.fif) does not conform to MNE naming conventions. All raw files should end with raw.fif, raw_sss.fif, raw_tsss.fif, raw.fif.gz, raw_sss.fif.gz, raw_tsss.fif.gz or _meg.fif
  raw_orig = reader[ext](**raw._init_kwargs)
    Read a total of 8 projection items:
        grad_68.fif : PCA-v1 (1 x 306)  idle
        grad_68.fif : PCA-v2 (1 x 306)  idle
        grad_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v1 (1 x 306)  idle
        mag_68.fif : PCA-v2 (1 x 306)  idle
        mag_68.fif : PCA-v3 (1 x 306)  idle
        mag_68.fif : PCA-v4 (1 x 306)  idle
        mag_68.fif : PCA-v5 (1 x 306)  idle
    Range : 863000 ... 1633999 =    863.000 ...  1633.999 secs
Ready.
Traceback (most recent call last):
  File "/home/adina/env/magma/bin/pymento-restructure", line 33, in <module>
    sys.exit(load_entry_point('pymento-meg', 'console_scripts', 'pymento-restructure')())
  File "/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/cli.py", line 194, in restructure
    restructure_to_bids(rawdir=args.raw_data_dir,
  File "/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/pymento.py", line 40, in restructure_to_bids
    raw = read_data_original(directory=rawdir,
  File "/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/orig/restructure.py", line 101, in read_data_original
    save_bids_data(raw=raw,
  File "/home/adina/repos/PhD/memento-code/code/pymento/pymento_meg/orig/restructure.py", line 135, in save_bids_data
    write_raw_bids(raw, bids_path, events_data=events_data,
  File "/home/adina/repos/mne-bids/mne_bids/write.py", line 1149, in write_raw_bids
    assert_array_equal(raw.times, raw_orig.times,
  File "/home/adina/env/magma/lib/python3.9/site-packages/numpy/testing/_private/utils.py", line 932, in assert_array_equal
    assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
  File "/home/adina/env/magma/lib/python3.9/site-packages/numpy/testing/_private/utils.py", line 761, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal
raw.times should not have changed since reading in from the file. It may have been cropped.
(shapes (3827000,), (771000,) mismatch)
 x: array([0.000000e+00, 1.000000e-03, 2.000000e-03, ..., 3.826997e+03,
       3.826998e+03, 3.826999e+03])
 y: array([0.00000e+00, 1.00000e-03, 2.00000e-03, ..., 7.70997e+02,
       7.70998e+02, 7.70999e+02])
pymento-restructure --subject '005' --raw_data_dir  --bids_dir  --calfiles_di  3.99s user 1.78s system 146% cpu 3.941 total
```
</details>

### PR DESCRIPTION
**Pull request by hoechenberger**
This fails on `master`, but works with this PR:

```python
from pathlib import Path
from tempfile import NamedTemporaryFile

import mne
from mne.datasets import sample

from mne_bids import write_raw_bids,  BIDSPath

data_path = Path(sample.data_path())
raw_fname = data_path / 'MEG' /  'sample' / 'sample_audvis_raw.fif'
output_path = data_path / '..' /'MNE-sample-data-bids'
bids_path = BIDSPath(subject='01',  task='audiovisual', root=output_path)

raw = mne.io.read_raw_fif(raw_fname)
raw.info['line_freq'] = 60
raw_concat = mne.concatenate_raws([raw.copy(), raw])

write_raw_bids(raw=raw_concat, bids_path=bids_path, overwrite=True)
```

Addresses part of #718

What do you think?

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by sappelhoff** [comment]
this looks like an improvement to me, going from a very crude handling of the situation to detecting different cases and not raising in a case where an info suffices. :+1: 

btw, thanks @adswa for raising this in #718

**Comment by agramfort** [comment]
we are allowing to write derivative files here no? why is it two files in
the first place?

>

**Comment by hoechenberger** [comment]
> we are allowing to write derivative files here no? why is it two files in the first place?

No it's not a derivative; it's trying to solve the issue @adswa, who wanted to convert some non-BIDS data to BIDS, and concatenated data before calling `write_raw_bids()`. This doesn't work on `master`.

And yes, you **could** abuse the functionality here to also secretly create derivatives of concatenated BIDS data. But I don't think we can prevent that…

**Comment by agramfort** [comment]
the two raw files should not be considered as 2 different runs?

>

**Comment by jasmainak** [comment]
I think the "correct" solution is to fix this explicitly for split fif files ... if I'm understanding it properly?

**Comment by agramfort** [comment]
I don't think it's related to split fif files? they should be handled
transparently
unless they have been renamed manually

>

**Comment by hoechenberger** [comment]
> the two raw files should not be considered as 2 different runs?

Yes this is my understanding. It seems the originals are split files either **have** been renamed, or somebody simply created multiple files manually during the recording. You could argue that in the latter case, these are different runs, then.

@adswa, do you know which of the two situations you're facing? :)

**Comment by adswa** [comment]
In my case, the files are split automatically due to their size (a roughly 60min recording), not separate runs. :) I can't tell you how they got their inconsistent filenames, though...

**Comment by agramfort** [comment]
ok then I think you should rewrite them to disk with fixed names and then
use mne-bids.
I would not amend mne-bids. It's pb of the data.

my 2c

>

**Comment by adswa** [comment]
>  It's pb of the data

That is true, but, and I don't intend this to sound snarky in written form, the reason for me to use mne-bids is to fix this problem of the data. 

Personally, I agree to @sappelhoff's statement from above:

> from a very crude handling of the situation to detecting different cases

You have said

> I would not amend mne-bids

and that's a fair standpoint. My 2ct as how to this proposed amendment appears to me: It is minor, without API change, it is not an additional feature that would require maintenance. If I am seeing the situation correctly, the check as it currently is implemented would prevent renaming misnamed split files using mne-bids (which, for me, is the only reason I am using it in my scripts). The change in this PR would allow doing it in cases that appear like legit use cases, while still warning the user that they are shooting themselves potentially in the foot yet not classifying *every* situation as a gun pointing to a foot.

Independent from the change from error to warning for a subset of cases, the changes 

```py
    if (len(raw.times) == len(raw_orig.times) and
            not np.array_equal(raw.times, raw_orig.times)):
        raise ValueError(
            "raw.times should has changed since reading from the file, but "
            "write_raw_bids() doesn't allow writing modified files.")

    elif len(raw.times) < len(raw_orig.times):
        raise ValueError(
            "The raw data you want to write contains fewer time points than "
            "the raw data on disk. It is possible that you cropped your data, "
            "which write_raw_bids() won't accept.")
```

add more details to the previous check (have I cropped my data, or have I modified it in place?), which are useful assessments in the case that the check fails (imagine my confusion about a message about a potentially cropped file, when the file I am saving has actually more time points than the file it is checked against).

**Comment by hoechenberger** [comment]
I agree with both of you, all arguments are totally valid. 

If we agree that one of the goals of MNE-BIDS should also be to make it easy and to dort if incentivize converting existing messy datasets to BIDS, then I think we have two main options:

- apply the change proposed here, or
- amend our existing docs/tutorials to include the workaround I suggested in the original issue, i.e. writing the day to a temporary file before writing to BIDS

Thoughts?

**Comment by sappelhoff** [comment]
+1 for changes proposed here, although I find the "workaround" good enough as well in case somebody is totally against the changes here.

**Comment by agramfort** [comment]
if it was me alone I would do the work around. This fix introduces something ad-hoc
which I don't feel I could explain / justify in the future to users.

**Comment by hoechenberger** [comment]
I too now have the gut feeling that at least my current approach to the problem in this PR might be opening a can of worms... but I cannot really explain why.

I'd currently cautiously vote for a proper documentation of our workaround... although I'm not entirely sure how to document it best

😬🤯

**Comment by jasmainak** [comment]
I would change this PR to remove the logger.info and keep the ValueErrors. As in it should still throw an error but no objection if you want to make it more informative.

@adswa Regarding split file names, if you try to read the file in MNE, doesn't it tell you what is the [next file name it expects](https://github.com/mne-tools/mne-python/blob/main/mne/io/fiff/raw.py#L89)? You can then use that information to rename the file. If there are too many of them, you can do a try/catch block for the exception and extracts the filename for the error message. Once you have renamed the file, you can use MNE-BIDS to BIDSify the data. Does that workflow make sense to you?

File: mne_bids/write.py [hunk 1]
```diff
@@ -1146,9 +1146,25 @@ def write_raw_bids(raw, bids_path, events_data=None,
...
+            "the raw data on disk. It is possible that you cropped your data, "
+            "which write_raw_bids() won't accept.")
+    elif len(raw.times) > len(raw_orig.times):
+        if np.array_equal(raw.times[:len(raw_orig.times)], raw_orig.times):
+            logger.info('Raw object contains more time points than the '
+                        'original file on disk; assuming concatenated data.')
+        else:
+            raise ValueError(
+                "Raw object contains more time points than the original file "
+                "on disk; but the first time points don't match.")
```
Comment (jasmainak) [inline mne_bids/write.py:1166]: also ... you will reduce code coverage this way @hoechenberger . Need to add more tests ... no easy solutions when I'm reviewing ;-)

Comment (hoechenberger) [reply mne_bids/write.py:1166]: I know, I just wanted to get some comments in this proof of concept before adding tests :)

**Comment by adswa** [comment]
> @adswa Regarding split file names, if you try to read the file in MNE, doesn't it tell you what is the next file name it expects? You can then use that information to rename the file. If there are too many of them, you can do a try/catch block for the exception and extracts the filename for the error message. Once you have renamed the file, you can use MNE-BIDS to BIDSify the data. Does that workflow make sense to you?

Makes perfect sense, thanks. I won't be renaming the file in-place though, there are too many other people that have based their work on them. I will either use mne's ``.save(split_naming=‘bids’)`` on the concatenated file, which works nicely at all times, or stick to the ``with NamedTemporaryFile`` solution @hoechenberger kindly pointed to in the original issue. :)

**Comment by hoechenberger** [comment]
So I've modified things such that we now always raise an exception when we encounter altered `raw.times`, but we provide a more detailed error message, also suggesting that users may reach out to us if they believe their use case should be supported. WDYT?

**Review by sappelhoff — APPROVED** [review]
small but nice improvement, and we can wait to see if this becomes a common complaint.

If yes, we can revisit and continue thinking about it / acting upon it.

**Comment by adswa** [comment]
Just for the record, to "reach out to the mne-developers", I do believe I have a valid use case that warrants support, but I think this PR is an improvement to the previous situation, nevertheless. thanks :) 

> I'd currently cautiously vote for a proper documentation of our workaround... 

Documentation for the workaround would be great, though.

**Comment by hoechenberger** [comment]
@adswa 
I'm worried that if we add this to the official docs, some users will start working around this limitation in cases where it's not warranted. My idea was to encourage them to read out so we can discuss things on an individual level, as we did with you.

Is this unsatisfactory..?

**Comment by hoechenberger** [comment]
Ugh I have no idea what's up with CI, it all passes flawlessly locally for me

**Comment by adswa** [comment]
> I'm worried that if we add this to the official docs, some users will start working around this limitation in cases where it's not warranted. My idea was to encourage them to read out so we can discuss things on an individual level, as we did with you.

Ah, I don't want to pester you. My personal opinion is that there only very rarely is "too much" documentation as long as it means something to someone, and  Sphinx has fantastic features to mark up aspects of documentation as not being part of the main code to execute, but I mostly referred to your own [comment](https://github.com/mne-tools/mne-bids/pull/719#issuecomment-790738725).

**Review by agramfort — APPROVED** [review]
+1 to merge when CIs are green.

**Comment by hoechenberger** [comment]
> Ugh I have no idea what's up with CI, it all passes flawlessly locally for me

Ok seems what I'm doing in one test doesn't work with 0.22, only `main`. Will look into this later today.

**Comment by hoechenberger** [comment]
> but I mostly referred to your own [comment](https://github.com/mne-tools/mne-bids/pull/719#issuecomment-790738725).

Yes, my opinions sometimes oscillate quickly 😅

**Comment by hoechenberger** [comment]
@adswa

Ok I've added a tiny tutorial, WDYT?

https://4173-89170358-gh.circle-artifacts.com/0/dev/auto_examples/write_modified_files.html#sphx-glr-auto-examples-write-modified-files-py

**Comment by adswa** [comment]
Great! Thank you for taking the time to do this!

File: examples/write_modified_files.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+with NamedTemporaryFile(suffix='_raw.fif') as f:
+    fname = f.name
+    raw_concat.save(fname, overwrite=True)
+    raw_concat = mne.io.read_raw_fif(fname, preload=False)
+    write_raw_bids(raw=raw_concat, bids_path=bids_path, overwrite=True)
+
+###############################################################################
+# That's it!
+#
+# .. warning:: **Remmeber, this should only ever be a last resort!**
```
Comment (adswa) [inline examples/write_modified_files.py:76]:
```suggestion
# .. warning:: **Remember, this should only ever be a last resort!**
```

I caught a typo (marked as a suggestion for better readability), maybe you want to change it yourself and force-push.

Comment (hoechenberger) [reply examples/write_modified_files.py]: thanks!

**Comment by hoechenberger** [comment]
Tutorial doc artifact:
https://4176-89170358-gh.circle-artifacts.com/0/dev/auto_examples/write_modified_files.html

**Review by sappelhoff — APPROVED** [review]
I'll let alex merge when he is happy

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
thx @hoechenberger

**Pull request merged**